transport/src/main/java/io/netty/channel/socket/nio/NioServerSocketChannel.java

```diff
@@ -137,18 +152,34 @@ protected void doClose() throws Exception {
 
     @Override
     protected int doReadMessages(List<Object> buf) throws Exception {
-        SocketChannel ch = javaChannel().accept();
+        if (connectionCounter.incrementAndGet() > maxConnections) {
+            connectionCounter.decrementAndGet();
+            try {
+                //sleep a while so that we won't loop busily
+                Thread.sleep(10);
+            }catch (InterruptedException exception){
+                //ignore this
+            }
+            return 0;
+        }
 
+        SocketChannel ch = null;
         try {
+            ch = javaChannel().accept();
             if (ch != null) {
-                buf.add(new NioSocketChannel(this, ch));
+                NioSocketChannel nioSocketChannel = new NioSocketChannel(this, ch);
+                nioSocketChannel.closeFuture().addListener(closeListener);
```

Comment:
u are right.   we  may save the listener if the maxconnection feature is not turned on ;)